lib/newProfile.ml

```diff
@@ -0,0 +1,204 @@
+(************************************************************************)
+(*         *   The Coq Proof Assistant / The Coq Development Team       *)
+(*  v      *         Copyright INRIA, CNRS and contributors             *)
+(* <O___,, * (see version control and CREDITS file for authors & dates) *)
+(*   \VV/  **************************************************************)
+(*    //   *    This file is distributed under the terms of the         *)
+(*         *     GNU Lesser General Public License Version 2.1          *)
+(*         *     (see LICENSE file for the text of the license)         *)
+(************************************************************************)
+
+module MiniJson = struct
+  type t =[
+    | `Intlit of string
+    | `String of string
+    | `Assoc of (string * t) list
+    | `List of t list
+  ]
+
+  let rec pr ch : t -> _ = function
+    | `String s ->
+      let s = String.split_on_char '"' s in
+      let s = String.concat "\\\"" s in
+      Format.fprintf ch "\"%s\"" s
+    | `Intlit s -> Format.fprintf ch "%s" s
+    | `Assoc elts ->
+      Format.fprintf ch "{ %a }" prrecord elts
+    | `List elts ->
+      Format.fprintf ch "[\n%a\n]" prarray elts
+
+  and prrecord ch = function
+    | [] -> ()
+    | [(x,v)] -> Format.fprintf ch "\"%s\": %a" x pr v
+    | (x,v) :: l -> Format.fprintf ch "\"%s\": %a, %a" x pr v prrecord l
+
+  and prarray ch = function
+    | [] -> ()
+    | [x] -> pr ch x
+    | x :: l -> Format.fprintf ch "%a,\n%a" pr x prarray l
+
+
+  let pid = Unix.getpid()
+
+  let pids = string_of_int pid
+  let base = [("pid", `Intlit pids); ("tid", `Intlit pids)]
+
+  let duration ~name ~ph ~ts ?args () =
+    let l = ("name", `String name) :: ("ph", `String ph) :: ("ts", `Intlit ts) :: base in
+    let l = match args with
+      | None -> l
+      | Some args -> ("args", `Assoc args) :: l
+    in
+    `Assoc l
+end
+
+type accu = {
+  ch : Format.formatter;
+  mutable sums : (float * (float * int) CString.Map.t) list;
+}
+
+let accu = ref None
+
+let is_profiling () = Option.has_some !accu
+
+let f fmt = match !accu with
+  | None -> assert false
+  | Some { ch } -> Format.fprintf ch fmt
```

Comment:
@SkySkimmer you may want to add a flush here if you want to have fully synchonous semantics, `Format` can buffer aggresively some times.

This can also be done using the specifier `%!` , or via the `kprintf` family of functions.